server/src/main/java/org/apache/uniffle/server/ShuffleServerGrpcService.java

```diff
@@ -256,6 +251,9 @@ public void sendShuffleData(SendShuffleDataRequest req,
           break;
         }
       }
+      // since the required buffer id is only used once, the shuffle client would try to require another buffer.
+      // we should remove this require buffer whether the cache shuffle data succeeded or not
+      manager.removeRequireBufferId(requireBufferId);
```

Comment:
We can't remove it after all blocks cached.

https://github.com/apache/incubator-uniffle/blob/c8f8292ce03d4a0f481759d79e404a3324869612/server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java#L497-L514

Reply:
Another reason:  Pre allocated memory cannot be released in time

Reply:
If all goes well, preAllocated memory will be released when cache block.  but `usedMemory` will be released after flush finish. And `preAllocatedBufferCheck` will release `usedMemory` 
So, if preAllocatedBufferCheck clean the require buffer when caching blocks, `usedMemory` will be released twice.